### PR TITLE
optional-component secure-data-proxy + related configs in cowbird/magpie/weaver

### PR DESCRIPTION
**Pull request by fmigneault**
Multiple updates to support secured WPS outputs.

## Changes

**Non-breaking changes**
- secure-data-proxy: add new [`secure-data-proxy`][secure-data-proxy] optional component.

  When enabled, this component will enforce authentication and authorization to be resolved against the `/wpsoutputs`
  endpoint prior to accessing the results produced by WPS executions. A Magpie service named `secure-data-proxy` is
  created to define the resource and permission hierarchy of directories and files the users and groups can access.
  When disabled, the original behavior to provide open access to `/wpsoutputs` is employed.

  A variable named `SECURE_DATA_PROXY_AUTH_INCLUDE` is dynamically assigned based on the activation or not of this
  component. Corresponding validation of optional/mandatory/delayed-eval variables used by this component are also
  applied dynamically, as well as mounting the necessary `nginx` and `docker-compose` extended configurations.

- Weaver: adjust user-context output directory hooks and permissions for [`secure-data-proxy`][secure-data-proxy].

  When a process defined in Weaver (either a WPS provider or a local definition) is executed by a user that was granted
  authorization to run a job, the corresponding user-context directory under `/wpsoutputs/users/{user-id}` will be used
  for storing the execution outputs and will have the appropriate permissions set for that user to grant them access to
  those outputs.

- Magpie/Twitcher: update minimum version `magpie>=3.31.0` to employ `twitcher>=0.8.0` in `MapgieAdatepr`.

  - Resolve an issue where `response.request` references were not set in OWS proxy responses when handled by Twitcher.
    This caused `MapgieAdatepr` response hooks to fail, which in turn caused failing requests for any non-WPS
    service that defined any proxy request hook, such as in the case of [`weaver`][weaver-component] component.

  - Adds the Twitcher ``/ows/verify/{service_name}[/{extra_path}`` endpoint employed for validating authorized access
    to Magpie service/resources, in the same fashion as the protected proxy endpoint, but without performing the proxied
    request toward the target service. This is mandatory for using the new [`secure-data-proxy`][secure-data-proxy] 
    optional component, otherwise the proxy endpoint triggers data download twice, once for authorization and another
    for actually accessing the data.

**Breaking changes**
- n/a (technically)
  New feature  [`secure-data-proxy`][secure-data-proxy] with optional component should default to using the original methodology of public access if not enabled.

## Related Issue / Discussion

- Depends on https://github.com/Ouranosinc/Magpie/pull/571, which in turn depends on https://github.com/bird-house/twitcher/pull/118. Corresponding Dockers `magpie>=3.31.0` and `twitcher>=0.8.0` must be deployed and used.
- Resolves [DAC-469 - *Secure WPS-Outputs served by Nginx*](https://crim-ca.atlassian.net/browse/DAC-469)
- Relates to [DAC-145 - *2.5.1 - Create private directory for new user for climate data [Cowbird/Thredds]*](https://crim-ca.atlassian.net/browse/DAC-145)
  (planing ahead some related permissions to user-context / wps outputs under weaver)

## Testing

To test the feature, simply enabled it in the `EXTRA_CONF_DIRS`, and place some files under `${DATA_PERSIST_ROOT}/wpsoutputs`. 
There should be a predefined `/wpsoutputs/public` location where files would be accessible when not logged in. Other locations should return 403. Then, a specific directory/file can be defined in Magpie with the relevant user permission on that resource to validate access to that file when logged in.

## To Do

- [x] Review and deploy Twitcher from https://github.com/bird-house/twitcher/pull/118
- [x] Review and deploy Magpie from https://github.com/Ouranosinc/Magpie/pull/571
- [ ] Merge DACCS-config with `secure-data-proxy` branch for PR test suite

[secure-data-proxy]: birdhouse/optional-components/secure-data-proxy
[weaver-component]: birdhouse/components/weaver

### CODE REVIEW

File: birdhouse/config/proxy/conf.d/all-services.include.template [hunk 1]
```diff
@@ -39,18 +39,20 @@
     }
 
     location /wpsoutputs/ {
+        ${SECURE_DATA_PROXY_AUTH_INCLUDE}
```
[the remark this answers is not in the document]
Comment (fmigneault) [reply birdhouse/config/proxy/conf.d/all-services.include.template:42]: This should be resolved with ca829ec (tested locally). Test suite should eventually post results on this PR.

**Comment by fmigneault** [comment]
2 builds are expected after this comment ⬇️ (comments by test suite itself to appear eventually...) 
*look for the value of `DACCS_CONFIGS_BRANCH` in the posted comments*

- [ ] 1 without `secure-data-proxy` (using `DACCS_CONFIGS_BRANCH=master`)
- [ ] 1 with `secure-data-proxy` enabled (using `DACCS_CONFIGS_BRANCH=secure-data-proxy`)

**Comment by fmigneault** [comment]
@tlvu 
Let me know when you are done testing this on your end. I'm ready to integrate it. 
Also, so we can sync versions with other pending PRs.

File: birdhouse/default.env [hunk 2]
```diff
@@ -89,6 +91,16 @@ export MAGPIE_USER_REGISTRATION_NOTIFY_ENABLED=false
...
 
+# Endpoint to verify Magpie/Twitcher authorization to a service/resource by a user without proxy request
+# Requires Twitcher>=0.8.0, Required for 'optional-compontents/secure-data-proxy'
+# Uses single quotes for DELAYED_EVAL mechanic to infer the default location based on the expected proxy path.
+export TWITCHER_VERIFY_PATH='$(echo "${TWITCHER_PROTECTED_PATH}" | sed "s/proxy/verify/")'
+
+# When 'optional-compontents/secure-data-proxy' is not enabled, this ensures the variable is defined for nginx.
+# Otherwise, it raises that it is not defined. We want it to be empty when this component is not used.
+# When the component is enabled, the value should be defined by its own 'default.env' file.
+export SECURE_DATA_PROXY_AUTH_INCLUDE=""
```
Comment (tlvu) [inline birdhouse/default.env:102]: Variable from `optional-compontents/secure-data-proxy` leaking to the main `default.env`?  There is no way to not have this?

File: birdhouse/components/cowbird/default.env [hunk 3]
```diff
@@ -4,9 +4,9 @@
 
 # All env in this default.env can be overridden by env.local.
 
-# All env in this default.env must NOT depend on any other env.  If they do,
-# must use single quote to avoid early expansion before overrides in local.env
-# are applied and must add to the list of DELAYED_EVAL.
+# All env in this default.env must NOT depend on any other env.  If they do, they
+# must use single quotes to avoid early expansion before overrides in env.local
```
Comment (tlvu) [inline birdhouse/components/cowbird/default.env:8]: Woops, `local.env`, not sure what I was smoking at that moment, thanks for all the fixes.

File: birdhouse/pavics-compose.sh [hunk 4]
```diff
@@ -87,6 +88,7 @@ OPTIONAL_VARS='
   $AUTODEPLOY_EXTRA_SCHEDULER_JOBS
   $PROXY_READ_TIMEOUT_VALUE
   $PROXY_ROOT_LOCATION
+  $SECURE_DATA_PROXY_AUTH_INCLUDE
```
Comment (tlvu) [inline birdhouse/pavics-compose.sh:91]: This should not be here since it is dynamically added by `optional-components/secure-data-proxy/`

File: birdhouse/scripts/get-components-json.sh [hunk 5]
```diff
@@ -0,0 +1,41 @@
+#!/bin/sh
```
Comment (tlvu) [inline birdhouse/scripts/get-components-json.sh:1]: This is odd, why is this file here?  It's part of the other PR already merged!

File: birdhouse/config/proxy/conf.d/all-services.include.template (hunk 1 above)
[the remark this answers is not in the document]
Comment (tlvu) [reply birdhouse/config/proxy/conf.d/all-services.include.template:42]: Ah I see why you had to add `SECURE_DATA_PROXY_AUTH_INCLUDE` to the main `default.env`.  This is annoying.

File: birdhouse/default.env [hunk 6]
```diff
@@ -70,7 +72,7 @@ export CANARIE_MONITORING_EXTRA_CONF_DIR="/conf.d"
 export THREDDS_ORGANIZATION="Birdhouse"
 
 # Tag version that will be used to update Magpie API, Magpie CLI, and matching Twitcher with Magpie Adapter
-export MAGPIE_VERSION=3.26.0
+export MAGPIE_VERSION=3.31.0
```
Comment (tlvu) [inline birdhouse/default.env:75]: Ouch, I have to test Magpie DB upgrade.

**Review by tlvu — COMMENTED** [review]
PR looks okay for me since Cowbird and this new secure-data-proxy is not yet activated on Ouranos side, it's not critical that I test everything.

But since you bumped Magpie, I'll have to test DB upgrade.  I'll get back to you once done.

File: birdhouse/pavics-compose.sh (hunk 4 above)
Comment (tlvu) [reply birdhouse/pavics-compose.sh:91]: I see why, because you had to modify `proxy/conf.d/all-services.include.template`.

**Review by tlvu — APPROVED** [review]
Magpie DB upgrade seems to be fine, only 1 DB upgrade:
```
[2023-02-10 01:29:43,140] INFO       [MainThread][alembic.runtime.migration] Running upgrade 0c6269f410cd -> 5e5acc33adce, Case Insensitive Email Constraint
```

However this new Magpie version seems to have a bunch of warnings, anything to be worried about?
```
/opt/local/src/magpie/magpie/models.py:935: SAWarning: Enum 'length' argument is currently ignored unless native_enum is specified as False, including for DDL that renders VARCHAR in any case.  This may change in a future release.
  operation = sa.Column(sa.Enum(TokenOperation, name=TokenOperation.__name__, length=32), nullable=False)
/opt/local/src/magpie/magpie/db.py:44: SAWarning: relationship 'Group.users_dynamic' will copy column groups.id to column users_groups.group_id, which conflicts with relationship(s): 'Group.users' (copies groups.id to users_groups.group_id), 'User.groups' (copies groups.id to users_groups.group_id). If this
 is not the intention, consider if these relationships should be linked with back_populates, or if viewonly=True should be applied to one or more if they are read-only. For the less common case that foreign key constraints are partially overlapping, the orm.foreign() annotation can be used to isolate the co
lumns that should be written towards.   To silence this warning, add the parameter 'overlaps="groups,users"' to the 'Group.users_dynamic' relationship. (Background on this error at: https://sqlalche.me/e/14/qzyx)
  configure_mappers()
/opt/local/src/magpie/magpie/db.py:44: SAWarning: relationship 'Group.users_dynamic' will copy column users.id to column users_groups.user_id, which conflicts with relationship(s): 'Group.users' (copies users.id to users_groups.user_id), 'User.groups' (copies users.id to users_groups.user_id). If this is no
t the intention, consider if these relationships should be linked with back_populates, or if viewonly=True should be applied to one or more if they are read-only. For the less common case that foreign key constraints are partially overlapping, the orm.foreign() annotation can be used to isolate the columns
that should be written towards.   To silence this warning, add the parameter 'overlaps="groups,users"' to the 'Group.users_dynamic' relationship. (Background on this error at: https://sqlalche.me/e/14/qzyx)
  configure_mappers()
/opt/local/src/magpie/magpie/db.py:44: SAWarning: relationship 'Group.resources_dynamic' will copy column groups.id to column resources.owner_group_id, which conflicts with relationship(s): 'Group.resources' (copies groups.id to resources.owner_group_id), 'Resource.owner_group' (copies groups.id to resource
s.owner_group_id). If this is not the intention, consider if these relationships should be linked with back_populates, or if viewonly=True should be applied to one or more if they are read-only. For the less common case that foreign key constraints are partially overlapping, the orm.foreign() annotation can
 be used to isolate the columns that should be written towards.   To silence this warning, add the parameter 'overlaps="owner_group,resources"' to the 'Group.resources_dynamic' relationship. (Background on this error at: https://sqlalche.me/e/14/qzyx)
  configure_mappers()
/opt/local/src/magpie/magpie/db.py:44: SAWarning: relationship 'Resource.groups' will copy column resources.resource_id to column groups_resources_permissions.resource_id, which conflicts with relationship(s): 'Resource.group_permissions' (copies resources.resource_id to groups_resources_permissions.resourc
e_id). If this is not the intention, consider if these relationships should be linked with back_populates, or if viewonly=True should be applied to one or more if they are read-only. For the less common case that foreign key constraints are partially overlapping, the orm.foreign() annotation can be used to
isolate the columns that should be written towards.   To silence this warning, add the parameter 'overlaps="group_permissions"' to the 'Resource.groups' relationship. (Background on this error at: https://sqlalche.me/e/14/qzyx)
  configure_mappers()
/opt/local/src/magpie/magpie/db.py:44: SAWarning: relationship 'Resource.groups' will copy column groups.id to column groups_resources_permissions.group_id, which conflicts with relationship(s): 'Group.resource_permissions' (copies groups.id to groups_resources_permissions.group_id), 'GroupResourcePermissio
n.groups' (copies groups.id to groups_resources_permissions.group_id). If this is not the intention, consider if these relationships should be linked with back_populates, or if viewonly=True should be applied to one or more if they are read-only. For the less common case that foreign key constraints are par
tially overlapping, the orm.foreign() annotation can be used to isolate the columns that should be written towards.   To silence this warning, add the parameter 'overlaps="groups,resource_permissions"' to the 'Resource.groups' relationship. (Background on this error at: https://sqlalche.me/e/14/qzyx)
  configure_mappers()
/opt/local/src/magpie/magpie/db.py:44: SAWarning: relationship 'Resource.users' will copy column resources.resource_id to column users_resources_permissions.resource_id, which conflicts with relationship(s): 'Resource.user_permissions' (copies resources.resource_id to users_resources_permissions.resource_id
). If this is not the intention, consider if these relationships should be linked with back_populates, or if viewonly=True should be applied to one or more if they are read-only. For the less common case that foreign key constraints are partially overlapping, the orm.foreign() annotation can be used to isol
ate the columns that should be written towards.   To silence this warning, add the parameter 'overlaps="user_permissions"' to the 'Resource.users' relationship. (Background on this error at: https://sqlalche.me/e/14/qzyx)
  configure_mappers()
/opt/local/src/magpie/magpie/db.py:44: SAWarning: relationship 'User.groups_dynamic' will copy column users.id to column users_groups.user_id, which conflicts with relationship(s): 'Group.users' (copies users.id to users_groups.user_id), 'Group.users_dynamic' (copies users.id to users_groups.user_id), 'User
.groups' (copies users.id to users_groups.user_id). If this is not the intention, consider if these relationships should be linked with back_populates, or if viewonly=True should be applied to one or more if they are read-only. For the less common case that foreign key constraints are partially overlapping,
 the orm.foreign() annotation can be used to isolate the columns that should be written towards.   To silence this warning, add the parameter 'overlaps="groups,users,users_dynamic"' to the 'User.groups_dynamic' relationship. (Background on this error at: https://sqlalche.me/e/14/qzyx)
  configure_mappers()
/opt/local/src/magpie/magpie/db.py:44: SAWarning: relationship 'User.groups_dynamic' will copy column groups.id to column users_groups.group_id, which conflicts with relationship(s): 'Group.users' (copies groups.id to users_groups.group_id), 'Group.users_dynamic' (copies groups.id to users_groups.group_id),
 'User.groups' (copies groups.id to users_groups.group_id). If this is not the intention, consider if these relationships should be linked with back_populates, or if viewonly=True should be applied to one or more if they are read-only. For the less common case that foreign key constraints are partially over
lapping, the orm.foreign() annotation can be used to isolate the columns that should be written towards.   To silence this warning, add the parameter 'overlaps="groups,users,users_dynamic"' to the 'User.groups_dynamic' relationship. (Background on this error at: https://sqlalche.me/e/14/qzyx)
  configure_mappers()
/opt/local/src/magpie/magpie/db.py:44: SAWarning: relationship 'User.resource_permissions' will copy column users.id to column users_resources_permissions.user_id, which conflicts with relationship(s): 'Resource.users' (copies users.id to users_resources_permissions.user_id). If this is not the intention, c
onsider if these relationships should be linked with back_populates, or if viewonly=True should be applied to one or more if they are read-only. For the less common case that foreign key constraints are partially overlapping, the orm.foreign() annotation can be used to isolate the columns that should be wri
tten towards.   To silence this warning, add the parameter 'overlaps="users"' to the 'User.resource_permissions' relationship. (Background on this error at: https://sqlalche.me/e/14/qzyx)
  configure_mappers()

```

**Comment by tlvu** [comment]
Please remove `get-components-json.sh` from this PR, it should really not be here.

**Comment by tlvu** [comment]
> Please remove `get-components-json.sh` from this PR, it should really not be here.

Ah I see why, in `master` the file name has `include` in it !

File: birdhouse/default.env (hunk 2 above)
Comment (fmigneault) [reply birdhouse/default.env:102]: It must be there to resolve what @matprov flagged in https://github.com/bird-house/birdhouse-deploy/pull/283#discussion_r1099254990. 
This value must be set empty to remove this part: https://github.com/bird-house/birdhouse-deploy/blob/secure-data-proxy/birdhouse/config/proxy/conf.d/all-services.include.template#L42

File: birdhouse/components/cowbird/default.env (hunk 3 above)
Comment (fmigneault) [reply birdhouse/components/cowbird/default.env:8]: I did the same mistake multiple times.

File: birdhouse/pavics-compose.sh (hunk 4 above)
Comment (fmigneault) [reply birdhouse/pavics-compose.sh:91]: It must be specified even when not used by `optional-components/secure-data-proxy`, otherwise it is not substituted in https://github.com/bird-house/birdhouse-deploy/blob/secure-data-proxy/birdhouse/config/proxy/conf.d/all-services.include.template?rgh-link-date=2023-02-10T04%3A08%3A10Z#L42, which in turn causes nginx to complain about invalid syntax.

File: birdhouse/scripts/get-components-json.sh (hunk 5 above)
Comment (fmigneault) [reply birdhouse/scripts/get-components-json.sh:1]: Oh, left over item prior to rename (I cherry-picked it). Will remove.

**Comment by fmigneault** [comment]
> bunch of warnings, anything to be worried about?

Those are produced in preparation of SQLAalchemy 2.0. 
I'm waiting on https://github.com/ergo/ziggurat_foundations/issues/74 to resolve the warnings. 
In the meantime, `sqlalchemy<2` is pinned (https://github.com/Ouranosinc/Magpie/blob/3.31.0/requirements.txt#L48), so no issue.

File: birdhouse/default.env (hunk 6 above)
Comment (fmigneault) [reply birdhouse/default.env:75]: Ok as per https://github.com/bird-house/birdhouse-deploy/pull/283#pullrequestreview-1292276358

**Comment by fmigneault** [comment]
@tlvu Just so we are in sync, I'll bump this to 1.23.0 now and merge.
